modules/consensus/accept.go

```diff
@@ -82,51 +47,65 @@ func (s *State) validHeader(b types.Block) (err error) {
 		return ErrEarlyTimestamp
 	}
 
-	// Check that the block is not too far in the future. An external process
-	// will need to be responsible for resubmitting the block once it is no
-	// longer in the future.
-	if b.Timestamp > types.CurrentTimestamp()+types.FutureThreshold {
-		return ErrFutureTimestamp
+	// If the block is in the extreme future, return an error and do nothing
+	// more with the block. There is an assumption that by the time the extreme
+	// future arrives, this block will no longer be a part of the longest fork
+	// because it will have been ignored by all of the miners.
+	if b.Timestamp > types.CurrentTimestamp()+types.ExtremeFutureThreshold {
+		return ErrExtremeFutureTimestamp
 	}
 
 	// Verify that the miner payouts sum to the total amount of fees allowed to
 	// be collected by the miners.
-	err = s.checkMinerPayouts(b)
-	if err != nil {
-		return
+	if !b.CheckMinerPayouts(parent.height + 1) {
+		return ErrBadMinerPayouts
+	}
+
+	// If the block is in the near future, but too far to be acceptable, then
+	// the block will be saved and added to the consensus set after it is no
+	// longer too far in the future. This is the last check because it's an
+	// expensive check, and not worth performing if the payouts are incorrect.
+	if b.Timestamp > types.CurrentTimestamp()+types.FutureThreshold {
+		go func() {
+			time.Sleep(time.Duration(b.Timestamp-(types.CurrentTimestamp()+types.FutureThreshold)) * time.Second)
+			lockID := cs.mu.Lock()
+			defer cs.mu.Unlock(lockID)
+			cs.acceptBlock(b) // TODO: How does this error get handled?
+		}()
+		return ErrFutureTimestamp
 	}
 
-	return
+	return nil
 }
 
 // addBlockToTree inserts a block into the blockNode tree by adding it to its
 // parent's list of children. If the new blockNode is heavier than the current
 // node, the blockchain is forked.
-func (s *State) addBlockToTree(b types.Block) (err error) {
-	parentNode := s.blockMap[b.ParentID]
+func (cs *State) addBlockToTree(b types.Block) (revertedNodes, appliedNodes []*blockNode, err error) {
+	parentNode := cs.blockMap[b.ParentID]
 	newNode := parentNode.newChild(b)
 
 	// Add the node to the block map
-	s.blockMap[b.ID()] = newNode
+	cs.blockMap[b.ID()] = newNode
 
-	if newNode.heavierThan(s.currentBlockNode()) {
-		err = s.forkBlockchain(newNode)
+	if newNode.heavierThan(cs.currentBlockNode()) {
+		revertedNodes, appliedNodes, err = cs.forkBlockchain(newNode)
 		if err != nil {
-			return
+			return nil, nil, err
 		}
+		return revertedNodes, appliedNodes, nil
```

Comment:
I've become a big fan of explicitly declaring your intentions in your return statements. If you say `return nil` instead of `return err`, then you are explicitly saying "if we reach this code, err is supposed to be nil, and even if I forgot to reset it or something, I'm going to return nil".

I would imagine that it also helps reviewers understand what's supposed to be happening.